### DISCUSSION TITLE
Other examples?

### DISCUSSION THREAD

**Discussion by udif** [MikePopoloski/slang#605] [Q&A]
Now that slang is able to parse my whole design, I'm looking for more fully compiling, standalone tools, that are using slang, beyond the simple 2 simple examples in  `source/util`. I searched for external standalone projects using it, but couldn't find any.

Also, I was assuming slang contains an elaboration pass that does the following:
1. Traverses the tree from the top level module
2. Calculates all parameter values, and all constant expressions used for bus widths, `generate` elaborations, etc.
 I realize there are issues with defparam statements, that requires a solver, but for the time being, I would simply raise an error if we encounter one, as 99% of the designs in the last 20 years are not using it.
  It also needs to resolve constant functions (compile it for a VM??)
4. Iterates through all the `generate` conditionals and loops, producing 
The end result of such a pass would be a new tree of concrete instances that can be more easily queried for actual design parameters.

I've searched the codebase, but I couldn't find any such pass. Am I missing something? Has someone else already wrote such a pass?
I have some ideas for an efficient storage scheme, by adding pointers to arrays of concrete values to relevant tree entries.
In these arrays the elements are stored in sequence but represent all possible instances from all the tree, relying on the instance order created by a deterministic (depth-first?) tree traversing algorithm.
What I still need to work out is how to store "hints" in the tree allowing efficient and simple access to the specific index range within these arrays for any point in the concrete elaborated tree.

**Comment by MikePopoloski** [comment]
I know of some tools using slang as a frontend, but they aren't open source (and there are probably more I don't know about too). It'd be nice if you wanted to build one and make it open source!

Yes, slang already does full elaboration. If you have a Compilation object, calling getRoot() will give you the root of the elaborated design tree, which you can walk to figure out anything about the design that you might like. defparams are mostly supported; there are some edge cases not fully handled but if you just use them for basic purposes it should be fine.

**Reply by udif** [reply]
Thanks! I will try this. I want to write a tool that produces a fully elaborated listing of the parametrized parts of a design, e.g.:
(Listed according to priority):

1. Instance tree (exact list of module/generate instances and their names)
2. Actual parameter values of each module instance
3. Full flattened listing of the modules.

All my work is open source, and done outside of working hours or company equipment.

**Reply by MikePopoloski** [reply]
Yeah, this should all be easy. The --ast-json option will already output #1 and #2, though the output is json and so can be very verbose for large designs. #3 shouldn't be all that hard to do either.

**Reply by shalakar** [reply]
Hi,  what is the status of this project? I'm very interested in this, is there anything I can do to help?
